staging/src/k8s.io/apiserver/pkg/server/options/opentelemetry.go

```diff
@@ -0,0 +1,140 @@
+	ConfigFile string
+}
+
+// NewOpenTelemetryOptions creates a new instance of OpenTelemetryOptions
+func NewOpenTelemetryOptions() *OpenTelemetryOptions {
+	return &OpenTelemetryOptions{}
+}
+
+// AddFlags adds flags related to opentelemetry to the specified FlagSet
+func (o *OpenTelemetryOptions) AddFlags(fs *pflag.FlagSet) {
+	if o == nil {
+		return
+	}
+
+	fs.StringVar(&o.ConfigFile, "opentelemetry-config-file", o.ConfigFile,
+		"File with apiserver opentelemetry configuration.")
+}
+
+// Apply adds the opentelemetry settings to the global configuration.
+func (o *OpenTelemetryOptions) Apply(es *egressselector.EgressSelector) error {
+	if o == nil {
+		return nil
+	}
+
+	npConfig, err := opentelemetry.ReadOpenTelemetryConfiguration(o.ConfigFile)
+	if err != nil {
+		return fmt.Errorf("failed to read opentelemetry config: %v", err)
+	}
+	errs := opentelemetry.ValidateOpenTelemetryConfiguration(npConfig)
+	if len(errs) > 0 {
+		return fmt.Errorf("failed to validate opentelemetry configuration: %v", errs.ToAggregate())
+	}
+
+	if npConfig == nil {
+		// No config file was specified, so don't enable exporting
+		return nil
+	}
+
+	opts := []otlp.ExporterOption{}
+	if npConfig.URL != nil {
+		opts = append(opts, otlp.WithAddress(*npConfig.URL))
+
+		if es != nil {
+			// Only use the egressselector dialer if egressselector is enabled.
+			// URL is on the "ControlPlane" network
+			egressDialer, err := es.Lookup(egressselector.ControlPlane.AsNetworkContext())
+			if err != nil {
+				return err
+			}
+
+			otelDialer := func(ctx context.Context, addr string) (net.Conn, error) {
+				return egressDialer(ctx, "tcp", addr)
+			}
+			opts = append(opts, otlp.WithGRPCDialOption(grpc.WithContextDialer(otelDialer)))
+		}
+	}
+	if npConfig.Service != nil {
+		// Default port is 55680
+		port := int32(55680)
+		if npConfig.Service.Port != nil {
+			port = *npConfig.Service.Port
+		}
+		addr := fmt.Sprintf("%s.%s:%d", npConfig.Service.Name, npConfig.Service.Namespace, port)
+		opts = append(opts, otlp.WithAddress(addr))
+
+		if es != nil {
+			// Only use the egressselector dialer if egressselector is enabled.
+			// Note that if not using egress selectors, it will try and call
+			// out to the service on the master.  For this to work, services
+			// must be accessible from the master, and would require running
+			// kube-proxy or a similar implementation of kubernetes services.
+
+			// Service is on the "Cluster" newtork
+			egressDialer, err := es.Lookup(egressselector.Cluster.AsNetworkContext())
+			if err != nil {
+				return err
+			}
+
+			otelDialer := func(ctx context.Context, addr string) (net.Conn, error) {
+				return egressDialer(ctx, "tcp", addr)
+			}
+			opts = append(opts, otlp.WithGRPCDialOption(grpc.WithContextDialer(otelDialer)))
+		}
+	}
+
+	traces.InitTraces("kube-apiserver", opts...)
+	return nil
+}
+
+// Validate verifies flags passed to OpenTelemetryOptions.
+func (o *OpenTelemetryOptions) Validate() []error {
```

Comment:
`func (o *OpenTelemetryOptions) Validate() (errs []error) {`

I think allows you to drop line 133 (you'll have to remove the nil on 130 too)

Reply:
done